presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -147,38 +147,20 @@ public int compare(JoinEnumerationResult result1, JoinEnumerationResult result2)
             };
         }
 
-        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols, boolean isCrossJoinPhase)
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
         {
             Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
             JoinEnumerationResult bestResult = memo.get(multiJoinKey);
             if (bestResult == null) {
                 checkState(sources.size() > 1, "sources size is less than or equal to one");
-                List<Optional<JoinEnumerationResult>> results = generatePartitions(sources.size())
-                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning, isCrossJoinPhase))
-                        .collect(toImmutableList());
+                bestResult = generatePartitions(sources.size())
+                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning))
+                        .filter(result -> !result.cost.isUnknown())
+                        .min(resultOrdering).orElse(UNKNOWN_COST_RESULT);
 
-                // If costs are unknown we can't compare them. We return a result with unknown cost.
```

Comment:
I think we still want to return original plan if there is at least one `UNKNOWN_COST` in https://github.com/Teradata/presto/pull/584/commits/b6e1fa7eaf889f06f9fc1cde73630469662d6123#diff-5cdb22a9680530f837d520839d7eaf99R156.
Because of https://github.com/Teradata/presto/pull/584/commits/b6e1fa7eaf889f06f9fc1cde73630469662d6123#diff-5cdb22a9680530f837d520839d7eaf99R158 this is no longer the case.
That is why I opted for having `UNKNOWN_COST` (for plans for which cost cannot be computed) and `INFINITE_COST` (for cross joins). `UNKNOWN_COST` will cut computations, while `INFINITE_COST` will allow computations for finish (explore alternative partitionings) and will be ignored in https://github.com/Teradata/presto/pull/584/commits/b6e1fa7eaf889f06f9fc1cde73630469662d6123#diff-5cdb22a9680530f837d520839d7eaf99R105